File: ios/DesMoinesInsider/ViewModels/MapViewModel.swift
import Foundation
import MapKit
import SwiftUI

/// ViewModel for the map view showing nearby events, restaurants, and attractions.
///
/// Annotation arrays (`eventAnnotations`, `restaurantAnnotations`,
/// `attractionAnnotations`) are **stored** properties recomputed only when
/// source data or category toggles change. Previously these were computed
/// properties that allocated fresh arrays on every SwiftUI re-read — with
/// 100+ pins that caused noticeable allocation pressure and map jank.
@MainActor
@Observable
final class MapViewModel {
    var events: [Event] = [] {
        didSet { refreshEventAnnotations() }
    }
    var restaurants: [Restaurant] = [] {
        didSet { refreshRestaurantAnnotations() }
    }
    var attractions: [Attraction] = [] {
        didSet { refreshAttractionAnnotations() }
    }
    private(set) var isLoading = false
    private(set) var errorMessage: String?
    private(set) var hasLoadedOnce = false

    var showEvents = true {
        didSet { refreshEventAnnotations() }
    }
    var showRestaurants = true {
        didSet { refreshRestaurantAnnotations() }
    }
    var showAttractions = true {
        didSet { refreshAttractionAnnotations() }
    }
    var selectedEvent: Event?
    var selectedRestaurant: Restaurant?
    var selectedAttraction: Attraction?
    var searchText = ""

    // Stored annotation caches. `private(set)` so the view can read but only
    // the ViewModel can mutate via the refresh helpers below.
    private(set) var eventAnnotations: [EventAnnotation] = []
    private(set) var restaurantAnnotations: [RestaurantAnnotation] = []
    private(set) var attractionAnnotations: [AttractionAnnotation] = []

    /// Time slider value (IOS-DISCOVER-2026-004). nil = "now" (live view).
    /// When non-nil, eventAnnotations are filtered to events within ±2h of
    /// this time and restaurantAnnotations to those open at this time.
    var mapTime: Date? = nil {
        didSet {
            refreshEventAnnotations()
            refreshRestaurantAnnotations()
        }
    }

    var cameraPosition: MapCameraPosition = .region(MKCoordinateRegion(
        center: CLLocationCoordinate2D(latitude: Config.defaultLatitude, longitude: Config.defaultLongitude),
        span: MKCoordinateSpan(latitudeDelta: 0.15, longitudeDelta: 0.15)
    ))

    private var fetchTask: Task<Void, Never>?

    private let locationService = LocationService.shared
    private let eventsService = EventsService.shared
    private let restaurantsService = RestaurantsService.shared
    private let attractionsService = AttractionsService.shared

    /// Current center coordinates used for data fetching
    private var currentLatitude = Config.defaultLatitude
    private var currentLongitude = Config.defaultLongitude

    // MARK: - Load Data

    func loadNearbyContent() async {
        // Cancel any in-flight fetch
        fetchTask?.cancel()

        // In UI testing mode, skip location services entirely to avoid
        // permission dialogs and network delays on CI simulators.
        if Config.isUITesting {
            await loadDefaultArea()
            return
        }

        isLoading = true
        errorMessage = nil

        do {
            let location = try await locationService.getCurrentLocation()
            guard !Task.isCancelled else { return }
            currentLatitude = location.coordinate.latitude
            currentLongitude = location.coordinate.longitude

            // Center map on user
            cameraPosition = .region(MKCoordinateRegion(
                center: location.coordinate,
                span: MKCoordinateSpan(latitudeDelta: 0.12, longitudeDelta: 0.12)
            ))

            await fetchAllContent()
        } catch {
            guard !Task.isCancelled else { return }
            // Fall back to Des Moines center
            await loadDefaultArea()
        }

        isLoading = false
        hasLoadedOnce = true
    }

    func loadDefaultArea() async {
        isLoading = true
        currentLatitude = Config.defaultLatitude
        currentLongitude = Config.defaultLongitude

        await fetchAllContent()

        isLoading = false
        hasLoadedOnce = true
    }

    /// Fetch all content types concurrently so one failure doesn't block the others.
    private func fetchAllContent() async {
        let lat = currentLatitude
        let lng = currentLongitude

        // Fetch all types concurrently - if one fails, others still load
        async let fetchedEvents: [Event] = {
            do { return try await eventsService.fetchNearbyEvents(latitude: lat, longitude: lng) }
            catch { return [] }
        }()

        async let fetchedRestaurants: [Restaurant] = {
            do { return try await restaurantsService.fetchNearbyRestaurants(latitude: lat, longitude: lng) }
            catch { return [] }
        }()

        async let fetchedAttractions: [Attraction] = {
            do { return try await attractionsService.fetchNearbyAttractions(latitude: lat, longitude: lng) }
            catch { return [] }
        }()

        let (e, r, a) = await (fetchedEvents, fetchedRestaurants, fetchedAttractions)
        events = e
        restaurants = r
        attractions = a

        if events.isEmpty && restaurants.isEmpty && attractions.isEmpty {
            if !Config.isConfigured {
                errorMessage = "Unable to connect to the server. Please check your internet connection."
            } else {
                errorMessage = "No places found in this area. Try expanding your search."
            }
        } else {
            errorMessage = nil
        }
    }

    // MARK: - Search

    func search() async {
        guard !searchText.trimmingCharacters(in: .whitespaces).isEmpty else {
            // Clear search - reload nearby content
            await fetchAllContent()
            return
        }

        isLoading = true
        errorMessage = nil
        clearSelection()

        // Search events by text
        let searchedEvents: [Event]
        do {
            searchedEvents = try await eventsService.fuzzySearchEvents(query: searchText)
        } catch {
            searchedEvents = []
        }

        // Search restaurants by text
        let searchedRestaurants: [Restaurant]
        do {
            searchedRestaurants = try await restaurantsService.fuzzySearchRestaurants(query: searchText)
        } catch {
            searchedRestaurants = []
        }

        // Search attractions by text
        let searchedAttractions: [Attraction]
        do {
            let response = try await attractionsService.fetchAttractions(
                query: .init(searchText: searchText, limit: 50)
            )
            searchedAttractions = response.attractions
        } catch {
            searchedAttractions = []
        }

        events = searchedEvents
        restaurants = searchedRestaurants
        attractions = searchedAttractions

        // If we got results, fit the map to show them. Compute the bounding
        // region off the main actor so 100+ coordinates don't block UI.
        let allCoords = (events.compactMap(\.coordinate)
                         + restaurants.compactMap(\.coordinate)
                         + attractions.compactMap(\.coordinate))
        if let region = await Self.regionFitting(coordinates: allCoords) {
            cameraPosition = .region(region)
        }

        if events.isEmpty && restaurants.isEmpty && attractions.isEmpty {
            errorMessage = "No results found for \"\(searchText)\""
        }

        isLoading = false
    }

    // MARK: - Retry

    func retry() async {
        await loadNearbyContent()
    }

    // MARK: - Clear Selection

    func clearSelection() {
        selectedEvent = nil
        selectedRestaurant = nil
        selectedAttraction = nil
    }

    // MARK: - Annotation Refresh

    /// Cached ISO parsers — `refreshEventAnnotations` runs on every map-time
    /// slider change and parses each event's timestamp, so re-allocating
    /// ISO8601DateFormatters per refresh (with 100+ pins) is wasteful.
    private static let isoFormatter: ISO8601DateFormatter = {
        let f = ISO8601DateFormatter()
        f.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
        return f
    }()
    private static let isoFormatterFallback: ISO8601DateFormatter = {
        let f = ISO8601DateFormatter()
        f.formatOptions = [.withInternetDateTime]
        return f
    }()

    private func refreshEventAnnotations() {
        guard showEvents else {
            if !eventAnnotations.isEmpty { eventAnnotations = [] }
            return
        }
        let timeWindow: (Date, Date)? = mapTime.map { t in
            (t.addingTimeInterval(-2 * 3600), t.addingTimeInterval(2 * 3600))
        }
        eventAnnotations = events.compactMap { event in
            guard let coord = event.coordinate else { return nil }
            // IOS-DISCOVER-2026-004: filter events to ±2h window when slider
            // is engaged. Prefer eventStartLocal, fall back to date.
            if let (lo, hi) = timeWindow {
                let candidates = [event.eventStartLocal, event.date].compactMap { $0 }
                let parsed = candidates.compactMap { s in
                    Self.isoFormatter.date(from: s) ?? Self.isoFormatterFallback.date(from: s)
                }
                guard let when = parsed.first, when >= lo && when <= hi else { return nil }
            }
            return EventAnnotation(event: event, coordinate: coord)
        }
    }

    private func refreshRestaurantAnnotations() {
        guard showRestaurants else {
            if !restaurantAnnotations.isEmpty { restaurantAnnotations = [] }
            return
        }
        let probeTime = mapTime
        restaurantAnnotations = restaurants.compactMap { restaurant in
            guard let coord = restaurant.coordinate else { return nil }
            // IOS-DISCOVER-2026-004: when slider is engaged, hide restaurants
            // that won't be open at the slider time. Best-effort hours parse —
            // restaurants without parseable hours are kept (rather than hidden)
            // so we don't drop a pin just because the source is missing data.
            if let probeTime, !Self.isLikelyOpen(restaurant: restaurant, at: probeTime) {
                return nil
            }
            return RestaurantAnnotation(restaurant: restaurant, coordinate: coord)
        }
    }

    /// Best-effort "is this restaurant open at the given time" check.
    /// Returns true when there's no businessHours data so the slider doesn't
    /// hide pins because of data gaps. Defers to Restaurant.isOpenNow(at:).
    private static func isLikelyOpen(restaurant: Restaurant, at time: Date) -> Bool {
        return restaurant.isOpenNow(at: time) ?? true
    }

    private func refreshAttractionAnnotations() {
        guard showAttractions else {
            if !attractionAnnotations.isEmpty { attractionAnnotations = [] }
            return
        }
        attractionAnnotations = attractions.compactMap { attraction in
            guard let coord = attraction.coordinate else { return nil }
            return AttractionAnnotation(attraction: attraction, coordinate: coord)
        }
    }

    var totalPinCount: Int {
        eventAnnotations.count + restaurantAnnotations.count + attractionAnnotations.count
    }

    // MARK: - Clustering

    /// Threshold above which we switch from per-pin rendering to grid clusters.
    /// Tuned for a 6.1" display at the default zoom (~0.12° span).
    static let clusterThreshold = 60

    /// Returns true when the total visible annotation count exceeds the
    /// threshold and clustering should be used instead of individual pins.
    var shouldCluster: Bool { totalPinCount > Self.clusterThreshold }

    /// Collapses every annotation into a uniform grid of `MapCluster` pins.
    /// Grid size scales with the visible span so clusters stay roughly
    /// the same size on screen as the user zooms.
    ///
    /// Callers should only evaluate this when `shouldCluster == true`; it is
    /// O(n) in the total annotation count and recomputes on each read.
    func clusters(for region: MKCoordinateRegion) -> [MapCluster] {
        // Bucket size ≈ 1/8th of the latitude span — tight enough to preserve
        // geographic meaning, coarse enough to collapse dense areas.
        let latBucket = max(region.span.latitudeDelta / 8.0, 0.005)
        let lngBucket = max(region.span.longitudeDelta / 8.0, 0.005)

        var buckets: [BucketKey: MapCluster] = [:]

        func ingest(id: String, coord: CLLocationCoordinate2D, kind: MapCluster.Kind) {
            let key = BucketKey(
                lat: Int((coord.latitude / latBucket).rounded()),
                lng: Int((coord.longitude / lngBucket).rounded())
            )
            if var existing = buckets[key] {
                existing.memberIds.append(id)
                existing.addCoordinate(coord)
                existing.kinds.insert(kind)
                buckets[key] = existing
            } else {
                buckets[key] = MapCluster(
                    id: "\(key.lat):\(key.lng)",
                    coordinate: coord,
                    memberIds: [id],
                    kinds: [kind]
                )
            }
        }

        for a in eventAnnotations { ingest(id: a.id, coord: a.coordinate, kind: .event) }
        for a in restaurantAnnotations { ingest(id: a.id, coord: a.coordinate, kind: .restaurant) }
        for a in attractionAnnotations { ingest(id: a.id, coord: a.coordinate, kind: .attraction) }

        return Array(buckets.values)
    }

    private struct BucketKey: Hashable {
        let lat: Int
        let lng: Int
    }

    var isEmpty: Bool {
        hasLoadedOnce && !isLoading && totalPinCount == 0
    }

    // MARK: - Helpers

    /// Compute a bounding region around the given coordinates. Runs on a
    /// detached task so large result sets (100+ pins) don't block the UI.
    nonisolated static func regionFitting(coordinates: [CLLocationCoordinate2D]) async -> MKCoordinateRegion? {
        guard !coordinates.isEmpty else { return nil }

        return await Task.detached(priority: .userInitiated) {
            var minLat = coordinates[0].latitude
            var maxLat = coordinates[0].latitude
            var minLng = coordinates[0].longitude
            var maxLng = coordinates[0].longitude

            for coord in coordinates {
                minLat = min(minLat, coord.latitude)
                maxLat = max(maxLat, coord.latitude)
                minLng = min(minLng, coord.longitude)
                maxLng = max(maxLng, coord.longitude)
            }

            let center = CLLocationCoordinate2D(
                latitude: (minLat + maxLat) / 2,
                longitude: (minLng + maxLng) / 2
            )
            let span = MKCoordinateSpan(
                latitudeDelta: max((maxLat - minLat) * 1.3, 0.02),
                longitudeDelta: max((maxLng - minLng) * 1.3, 0.02)
            )
            return MKCoordinateRegion(center: center, span: span)
        }.value
    }
}

// MARK: - Annotation Models

struct EventAnnotation: Identifiable {
    let event: Event
    let coordinate: CLLocationCoordinate2D
    var id: String { event.id }

    var tintColor: Color {
        guard let date = event.parsedDate else { return .blue }
        let calendar = Calendar.current
        if calendar.isDateInToday(date) { return .red }
        let days = calendar.dateComponents([.day], from: Date(), to: date).day ?? 0
        if days <= 7 { return .orange }
        return .blue
    }
}

struct RestaurantAnnotation: Identifiable {
    let restaurant: Restaurant
    let coordinate: CLLocationCoordinate2D
    var id: String { restaurant.id }
}

struct AttractionAnnotation: Identifiable {
    let attraction: Attraction
    let coordinate: CLLocationCoordinate2D
    var id: String { attraction.id }
}

// MARK: - Cluster

/// A grid-bucket aggregate used when the total annotation count exceeds
/// `MapViewModel.clusterThreshold`. Stores the centroid of its members and
/// which categories are represented so the pin view can tint accordingly.
struct MapCluster: Identifiable {
    enum Kind: Hashable {
        case event
        case restaurant
        case attraction
    }

    let id: String
    private(set) var coordinate: CLLocationCoordinate2D
    var memberIds: [String]
    var kinds: Set<Kind>

    var count: Int { memberIds.count }

    /// Update the stored coordinate to the running centroid after each member
    /// is added. Keeps the cluster pin visually anchored to its members.
    mutating func addCoordinate(_ coord: CLLocationCoordinate2D) {
        let n = Double(memberIds.count)
        let newLat = (coordinate.latitude * (n - 1) + coord.latitude) / n
        let newLng = (coordinate.longitude * (n - 1) + coord.longitude) / n
        coordinate = CLLocationCoordinate2D(latitude: newLat, longitude: newLng)
    }

    /// Primary tint for rendering: prefer events (red) > restaurants (orange) > attractions (green).
    var tintColor: Color {
        if kinds.contains(.event) { return .red }
        if kinds.contains(.restaurant) { return .orange }
        return .green
    }
}

File: android/app/src/main/java/com/desmoines/aipulse/ui/screens/map/MapViewModel.kt
package com.desmoines.aipulse.ui.screens.map

import android.Manifest
import android.content.Context
import android.content.pm.PackageManager
import android.util.Log
import androidx.core.content.ContextCompat
import com.desmoines.aipulse.data.model.Attraction
import com.desmoines.aipulse.data.model.Event
import com.desmoines.aipulse.data.model.Restaurant
import com.desmoines.aipulse.data.repository.AttractionsRepository
import com.desmoines.aipulse.data.repository.EventsRepository
import com.desmoines.aipulse.data.repository.RestaurantsRepository
import com.desmoines.aipulse.util.Config
import com.google.android.gms.location.FusedLocationProviderClient
import com.google.android.gms.location.LocationServices
import com.google.android.gms.location.Priority
import com.google.android.gms.maps.model.CameraPosition
import com.google.android.gms.maps.model.LatLng
import com.google.android.gms.tasks.CancellationTokenSource
import dagger.hilt.android.lifecycle.HiltViewModel
import dagger.hilt.android.qualifiers.ApplicationContext
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.combine
import kotlinx.coroutines.flow.SharingStarted
import kotlinx.coroutines.flow.stateIn
import kotlinx.coroutines.launch
import kotlinx.coroutines.suspendCancellableCoroutine
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import java.time.LocalDate
import java.time.OffsetDateTime
import java.time.temporal.ChronoUnit
import javax.inject.Inject
import kotlin.coroutines.resume
import kotlin.coroutines.resumeWithException

private const val TAG = "MapViewModel"

/**
 * Map screen state holder.
 * Mirrors iOS MapViewModel.swift.
 */
data class MapScreenState(
    val events: List<Event> = emptyList(),
    val restaurants: List<Restaurant> = emptyList(),
    val attractions: List<Attraction> = emptyList(),
    val showEvents: Boolean = true,
    val showRestaurants: Boolean = true,
    val showAttractions: Boolean = true,
    val selectedEvent: Event? = null,
    val selectedRestaurant: Restaurant? = null,
    val selectedAttraction: Attraction? = null,
    val searchText: String = "",
    val isLoading: Boolean = false,
    val errorMessage: String? = null,
    val hasLoadedOnce: Boolean = false,
    val cameraPosition: CameraPosition = CameraPosition.fromLatLngZoom(
        LatLng(Config.DEFAULT_LATITUDE, Config.DEFAULT_LONGITUDE), 12f
    ),
) {
    /** Event annotations filtered by showEvents toggle and valid coordinates. */
    val eventAnnotations: List<EventAnnotation>
        get() = if (!showEvents) emptyList() else events.mapNotNull { event ->
            event.coordinate?.let { coord ->
                EventAnnotation(
                    event = event,
                    position = LatLng(coord.latitude, coord.longitude)
                )
            }
        }

    /** Restaurant annotations filtered by showRestaurants toggle and valid coordinates. */
    val restaurantAnnotations: List<RestaurantAnnotation>
        get() = if (!showRestaurants) emptyList() else restaurants.mapNotNull { restaurant ->
            restaurant.coordinate?.let { coord ->
                RestaurantAnnotation(
                    restaurant = restaurant,
                    position = LatLng(coord.latitude, coord.longitude)
                )
            }
        }

    /** Attraction annotations filtered by showAttractions toggle and valid coordinates. */
    val attractionAnnotations: List<AttractionAnnotation>
        get() = if (!showAttractions) emptyList() else attractions.mapNotNull { attraction ->
            attraction.coordinate?.let { coord ->
                AttractionAnnotation(
                    attraction = attraction,
                    position = LatLng(coord.latitude, coord.longitude)
                )
            }
        }

    val totalPinCount: Int
        get() = eventAnnotations.size + restaurantAnnotations.size + attractionAnnotations.size

    val isEmpty: Boolean
        get() = hasLoadedOnce && !isLoading && totalPinCount == 0
}

// region Annotation Models

data class EventAnnotation(
    val event: Event,
    val position: LatLng,
) {
    val id: String get() = event.id

    /** Pin tint color: red=today, orange=≤7 days, blue=future. Matches iOS. */
    val urgencyColor: PinColor
        get() {
            val date = event.parsedDate ?: return PinColor.BLUE
            val now = OffsetDateTime.now()
            val today = now.toLocalDate()
            val eventDate = date.toLocalDate()
            if (eventDate == today) return PinColor.RED
            val daysUntil = ChronoUnit.DAYS.between(today, eventDate)
            if (daysUntil in 1..7) return PinColor.ORANGE
            return PinColor.BLUE
        }
}

data class RestaurantAnnotation(
    val restaurant: Restaurant,
    val position: LatLng,
) {
    val id: String get() = restaurant.id
}

data class AttractionAnnotation(
    val attraction: Attraction,
    val position: LatLng,
) {
    val id: String get() = attraction.id
}

enum class PinColor { RED, ORANGE, BLUE }

// endregion

@HiltViewModel
class MapViewModel @Inject constructor(
    @param:ApplicationContext private val appContext: Context,
    private val eventsRepository: EventsRepository,
    private val restaurantsRepository: RestaurantsRepository,
    private val attractionsRepository: AttractionsRepository,
) : ViewModel() {

    // Data
    private val _events = MutableStateFlow<List<Event>>(emptyList())
    private val _restaurants = MutableStateFlow<List<Restaurant>>(emptyList())
    private val _attractions = MutableStateFlow<List<Attraction>>(emptyList())

    // Toggles
    private val _showEvents = MutableStateFlow(true)
    private val _showRestaurants = MutableStateFlow(true)
    private val _showAttractions = MutableStateFlow(true)

    // Selection
    private val _selectedEvent = MutableStateFlow<Event?>(null)
    private val _selectedRestaurant = MutableStateFlow<Restaurant?>(null)
    private val _selectedAttraction = MutableStateFlow<Attraction?>(null)

    // Search
    private val _searchText = MutableStateFlow("")

    // State
    private val _isLoading = MutableStateFlow(false)
    private val _errorMessage = MutableStateFlow<String?>(null)
    private val _hasLoadedOnce = MutableStateFlow(false)
    private val _cameraPosition = MutableStateFlow(
        CameraPosition.fromLatLngZoom(
            LatLng(Config.DEFAULT_LATITUDE, Config.DEFAULT_LONGITUDE), 12f
        )
    )

    private var currentLatitude = Config.DEFAULT_LATITUDE
    private var currentLongitude = Config.DEFAULT_LONGITUDE

    private val fusedLocationClient: FusedLocationProviderClient =
        LocationServices.getFusedLocationProviderClient(appContext)

    val uiState: StateFlow<MapScreenState> = combine(
        combine(_events, _restaurants, _attractions, _showEvents, _showRestaurants) { events, restaurants, attractions, showE, showR ->
            MapScreenState(
                events = events,
                restaurants = restaurants,
                attractions = attractions,
                showEvents = showE,
                showRestaurants = showR,
            )
        },
        combine(_showAttractions, _selectedEvent, _selectedRestaurant, _selectedAttraction, _searchText) { showA, selE, selR, selA, search ->
            MapScreenState(
                showAttractions = showA,
                selectedEvent = selE,
                selectedRestaurant = selR,
                selectedAttraction = selA,
                searchText = search,
            )
        },
        combine(_isLoading, _errorMessage, _hasLoadedOnce, _cameraPosition) { loading, error, loaded, cam ->
            MapScreenState(
                isLoading = loading,
                errorMessage = error,
                hasLoadedOnce = loaded,
                cameraPosition = cam,
            )
        }
    ) { dataState, selectionState, uiStateChunk ->
        MapScreenState(
            events = dataState.events,
            restaurants = dataState.restaurants,
            attractions = dataState.attractions,
            showEvents = dataState.showEvents,
            showRestaurants = dataState.showRestaurants,
            showAttractions = selectionState.showAttractions,
            selectedEvent = selectionState.selectedEvent,
            selectedRestaurant = selectionState.selectedRestaurant,
            selectedAttraction = selectionState.selectedAttraction,
            searchText = selectionState.searchText,
            isLoading = uiStateChunk.isLoading,
            errorMessage = uiStateChunk.errorMessage,
            hasLoadedOnce = uiStateChunk.hasLoadedOnce,
            cameraPosition = uiStateChunk.cameraPosition,
        )
    }.stateIn(viewModelScope, SharingStarted.WhileSubscribed(5_000), MapScreenState())

    // region Public API

    fun loadNearbyContent() {
        viewModelScope.launch {
            _isLoading.value = true
            _errorMessage.value = null

            try {
                val location = getCurrentLocation()
                currentLatitude = location.latitude
                currentLongitude = location.longitude
                _cameraPosition.value = CameraPosition.fromLatLngZoom(location, 13f)
            } catch (e: Exception) {
                Log.d(TAG, "Location unavailable, using Des Moines default: ${e.message}")
                currentLatitude = Config.DEFAULT_LATITUDE
                currentLongitude = Config.DEFAULT_LONGITUDE
            }

            fetchAllContent()
            _isLoading.value = false
            _hasLoadedOnce.value = true
        }
    }

    fun search() {
        val query = _searchText.value.trim()
        if (query.isEmpty()) {
            // Clear search — reload nearby content
            viewModelScope.launch {
                _isLoading.value = true
                _errorMessage.value = null
                clearSelection()
                fetchAllContent()
                _isLoading.value = false
            }
            return
        }

        viewModelScope.launch {
            _isLoading.value = true
            _errorMessage.value = null
            clearSelection()

            // Search events by text (fuzzy)
            val searchedEvents = eventsRepository.fuzzySearchEvents(query)
                .getOrDefault(emptyList())

            // Search restaurants by text (fuzzy)
            val searchedRestaurants = restaurantsRepository.fuzzySearchRestaurants(query)
                .getOrDefault(emptyList())

            // Search attractions by text (ILIKE — no fuzzy RPC for attractions)
            val searchedAttractions = attractionsRepository.fetchAttractions(
                com.desmoines.aipulse.data.remote.AttractionsQuery(searchText = query, limit = 50)
            ).getOrNull()?.attractions ?: emptyList()

            _events.value = searchedEvents
            _restaurants.value = searchedRestaurants
            _attractions.value = searchedAttractions

            // Fit map to results with 30% padding
            val allCoords = buildList {
                searchedEvents.mapNotNull { it.coordinate }.forEach { add(LatLng(it.latitude, it.longitude)) }
                searchedRestaurants.mapNotNull { it.coordinate }.forEach { add(LatLng(it.latitude, it.longitude)) }
                searchedAttractions.mapNotNull { it.coordinate }.forEach { add(LatLng(it.latitude, it.longitude)) }
            }
            regionFitting(allCoords)?.let { _cameraPosition.value = it }

            if (searchedEvents.isEmpty() && searchedRestaurants.isEmpty() && searchedAttractions.isEmpty()) {
                _errorMessage.value = "No results found for \"$query\""
            }

            _isLoading.value = false
        }
    }

    fun retry() {
        loadNearbyContent()
    }

    fun setSearchText(text: String) {
        _searchText.value = text
    }

    fun toggleShowEvents() {
        _showEvents.value = !_showEvents.value
    }

    fun toggleShowRestaurants() {
        _showRestaurants.value = !_showRestaurants.value
    }

    fun toggleShowAttractions() {
        _showAttractions.value = !_showAttractions.value
    }

    fun selectEvent(event: Event) {
        clearSelection()
        _selectedEvent.value = event
    }

    fun selectRestaurant(restaurant: Restaurant) {
        clearSelection()
        _selectedRestaurant.value = restaurant
    }

    fun selectAttraction(attraction: Attraction) {
        clearSelection()
        _selectedAttraction.value = attraction
    }

    fun clearSelection() {
        _selectedEvent.value = null
        _selectedRestaurant.value = null
        _selectedAttraction.value = null
    }

    // endregion

    // region Private helpers

    /**
     * Fetch all content types independently — if one fails, others still load.
     * Matches iOS MapViewModel.fetchAllContent().
     */
    private suspend fun fetchAllContent() {
        val lat = currentLatitude
        val lng = currentLongitude

        val fetchedEvents = eventsRepository.fetchNearbyEvents(lat, lng)
            .getOrDefault(emptyList())

        val fetchedRestaurants = restaurantsRepository.fetchNearbyRestaurants(lat, lng)
            .getOrDefault(emptyList())

        val fetchedAttractions = attractionsRepository.fetchNearbyAttractions(lat, lng)
            .getOrDefault(emptyList())

        _events.value = fetchedEvents
        _restaurants.value = fetchedRestaurants
        _attractions.value = fetchedAttractions

        // Surface attractions (and events/restaurants) to on-device system search (ANDP-070).

        if (fetchedEvents.isEmpty() && fetchedRestaurants.isEmpty() && fetchedAttractions.isEmpty()) {
            _errorMessage.value = "No places found in this area. Try expanding your search."
        } else {
            _errorMessage.value = null
        }
    }

    /**
     * Get current device location using FusedLocationProviderClient.
     * Falls back to default if permission not granted.
     */
    private suspend fun getCurrentLocation(): LatLng {
        val hasPermission = ContextCompat.checkSelfPermission(
            appContext, Manifest.permission.ACCESS_FINE_LOCATION
        ) == PackageManager.PERMISSION_GRANTED || ContextCompat.checkSelfPermission(
            appContext, Manifest.permission.ACCESS_COARSE_LOCATION
        ) == PackageManager.PERMISSION_GRANTED

        if (!hasPermission) {
            throw SecurityException("Location permission not granted")
        }

        return suspendCancellableCoroutine { continuation ->
            val cancellationTokenSource = CancellationTokenSource()
            continuation.invokeOnCancellation { cancellationTokenSource.cancel() }

            fusedLocationClient.getCurrentLocation(
                Priority.PRIORITY_BALANCED_POWER_ACCURACY,
                cancellationTokenSource.token
            ).addOnSuccessListener { location ->
                if (location != null) {
                    continuation.resume(LatLng(location.latitude, location.longitude))
                } else {
                    continuation.resumeWithException(Exception("Location is null"))
                }
            }.addOnFailureListener { exception ->
                continuation.resumeWithException(exception)
            }
        }
    }

    /**
     * Calculate camera position that fits all coordinates with 30% padding.
     * Matches iOS MapViewModel.regionFitting().
     */
    private fun regionFitting(coordinates: List<LatLng>): CameraPosition? {
        if (coordinates.isEmpty()) return null

        var minLat = coordinates[0].latitude
        var maxLat = coordinates[0].latitude
        var minLng = coordinates[0].longitude
        var maxLng = coordinates[0].longitude

        for (coord in coordinates) {
            minLat = minOf(minLat, coord.latitude)
            maxLat = maxOf(maxLat, coord.latitude)
            minLng = minOf(minLng, coord.longitude)
            maxLng = maxOf(maxLng, coord.longitude)
        }

        val centerLat = (minLat + maxLat) / 2
        val centerLng = (minLng + maxLng) / 2
        val latSpan = maxOf((maxLat - minLat) * 1.3, 0.02)
        val lngSpan = maxOf((maxLng - minLng) * 1.3, 0.02)

        // Approximate zoom from span (rough heuristic — Google Maps doesn't use spans directly)
        val maxSpan = maxOf(latSpan, lngSpan)
        val zoom = when {
            maxSpan > 1.0 -> 9f
            maxSpan > 0.5 -> 10f
            maxSpan > 0.2 -> 11f
            maxSpan > 0.1 -> 12f
            maxSpan > 0.05 -> 13f
            else -> 14f
        }

        return CameraPosition.fromLatLngZoom(LatLng(centerLat, centerLng), zoom)
    }

    // endregion
}
